### PR TITLE
1992 - Add attribute references select dialog

### PR DESCRIPTION
**Pull request by orzechdev**
I want to merge this change because... it adds references select dialog to reference attributes.

Feature branch: https://github.com/mirumee/saleor-dashboard/pull/917

**PR intended to be tested with API branch:**  feature/page-reference-attributes
https://github.com/mirumee/saleor/pull/6624

### Screenshots

Zrzut ekranu 2021-01-5 o 13 12 35

### Pull Request Checklist

1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Translated strings are extracted.
1. [x] Number of API calls is optimized.
1. [x] The changes are tested.
1. [x] Data-test are added for new elements.
1. [x] Type definitions are up to date.
1. [ ] Changes are mentioned in the changelog.

### Test environment config

API_URI=https://feature-page-reference-attributes.api.saleor.rocks/graphql/

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-1992 Implement modal for selecting reference attribute values](https://app.clickup.com/t/2549495/SALEOR-1992)

File: src/attributes/utils/data.ts [hunk 1]
```diff
@@ -104,3 +121,38 @@ export const getAttributesAfterFileAttributesUpdate = (
...
 };
+
+export const getAttributesDisplayData = (
+  attributes: AttributeInput[],
+  attributesWithNewFileValue: FormsetData<null, File>,
+  referencePages: SearchPages_search_edges_node[]
+) =>
+  attributes.map(attribute => {
+    if (attribute.data.inputType === AttributeInputTypeEnum.REFERENCE) {
+      return {
```
Comment (mmarkusik) [inline src/attributes/utils/data.ts:132]: Would like to see content of this return as function `getParsedReferenceData` or something so this function is clearer in what it does

File: src/components/AssignAttributeValueDialog/AssignAttributeValueDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  const intl = useIntl();
+
+  return (
+    <AssignContainerDialog
+      containers={attributeValues.map(value => ({
+        id: value.id,
+        name: value.title
+      }))}
+      search={{
+        label: intl.formatMessage({
```
Comment (mmarkusik) [inline src/components/AssignAttributeValueDialog/AssignAttributeValueDialog.tsx:30]: Could look cleaner with those moved to defineMessages

File: src/pages/components/PageDetailsPage/PageDetailsPage.tsx [hunk 3]
```diff
@@ -161,6 +186,32 @@ const PageDetailsPage: React.FC<PageDetailsPageProps> = ({
...
+                data.attributes,
+                referencePages
+              )}
+              hasMore={fetchMoreReferencePages?.hasMore}
+              open={openAssignReferencesAttributeDialog}
+              onFetch={fetchReferncePages}
+              onFetchMore={fetchMoreReferencePages?.onFetchMore}
+              loading={fetchMoreReferencePages?.loading}
+              onClose={onCloseDialog}
+              onSubmit={attributeValues => {
```
Comment (mmarkusik) [inline src/pages/components/PageDetailsPage/PageDetailsPage.tsx:202]: Would look much cleaner with this moved outside of return

File: src/pages/views/PageCreate.tsx [hunk 4]
```diff
@@ -124,6 +135,27 @@ export const PageCreate: React.FC<PageCreateProps> = () => {
                 loading: searchPageTypesOpts.loading,
                 onFetchMore: loadMorePageTypes
               }}
+              assignReferencesAttributeId={
+                params.action === "assign-attribute-value" && params.id
+              }
+              onAssignReferencesClick={attribute => {
```
Comment (mmarkusik) [inline src/pages/views/PageCreate.tsx:141]: move outside of return

File: src/products/components/ProductUpdatePage/ProductUpdatePage.tsx [hunk 5]
```diff
@@ -362,6 +383,32 @@ export const ProductUpdatePage: React.FC<ProductUpdatePageProps> = ({
...
+                  data.attributes,
+                  referencePages
+                )}
+                hasMore={fetchMoreReferencePages?.hasMore}
+                open={openAssignReferencesAttributeDialog}
+                onFetch={fetchReferncePages}
+                onFetchMore={fetchMoreReferencePages?.onFetchMore}
+                loading={fetchMoreReferencePages?.loading}
+                onClose={onCloseDialog}
+                onSubmit={attributeValues => {
```
Comment (mmarkusik) [inline src/products/components/ProductUpdatePage/ProductUpdatePage.tsx:399]: same as above

File: src/products/components/ProductVariantCreatePage/ProductVariantCreatePage.tsx [hunk 6]
```diff
@@ -187,6 +209,32 @@ const ProductVariantCreatePage: React.FC<ProductVariantCreatePageProps> = ({
...
+                data.attributes,
+                referencePages
+              )}
+              hasMore={fetchMoreReferencePages?.hasMore}
+              open={openAssignReferencesAttributeDialog}
+              onFetch={fetchReferncePages}
+              onFetchMore={fetchMoreReferencePages?.onFetchMore}
+              loading={fetchMoreReferencePages?.loading}
+              onClose={onCloseDialog}
+              onSubmit={attributeValues => {
```
Comment (mmarkusik) [inline src/products/components/ProductVariantCreatePage/ProductVariantCreatePage.tsx:225]: same as above

File: src/products/components/ProductVariantPage/ProductVariantPage.tsx [hunk 7]
```diff
@@ -248,6 +271,32 @@ const ProductVariantPage: React.FC<ProductVariantPageProps> = ({
...
+                    data.attributes,
+                    referencePages
+                  )}
+                  hasMore={fetchMoreReferencePages?.hasMore}
+                  open={openAssignReferencesAttributeDialog}
+                  onFetch={fetchReferncePages}
+                  onFetchMore={fetchMoreReferencePages?.onFetchMore}
+                  loading={fetchMoreReferencePages?.loading}
+                  onClose={onCloseDialog}
+                  onSubmit={attributeValues => {
```
Comment (mmarkusik) [inline src/products/components/ProductVariantPage/ProductVariantPage.tsx:287]: +1

File: src/products/views/ProductCreate/ProductCreate.tsx [hunk 8]
```diff
@@ -267,6 +272,27 @@ export const ProductCreateView: React.FC<ProductCreateProps> = ({ params }) => {
         weightUnit={shop?.defaultWeightUnit}
         openChannelsModal={handleChannelsModalOpen}
         onChannelsChange={setCurrentChannels}
+        assignReferencesAttributeId={
+          params.action === "assign-attribute-value" && params.id
+        }
+        onAssignReferencesClick={attribute => {
```
Comment (mmarkusik) [inline src/products/views/ProductCreate/ProductCreate.tsx:278]: same as above

**Review by mmarkusik — CHANGES_REQUESTED** [review]

File: src/products/components/ProductUpdatePage/ProductUpdatePage.tsx [hunk 9]
```diff
@@ -169,6 +188,24 @@ export const ProductUpdatePage: React.FC<ProductUpdatePageProps> = ({
       value: taxType.taxCode
     })) || [];
 
+  const openAssignReferencesAttributeDialog = !!assignReferencesAttributeId;
```
Comment (dominik-zeglen) [inline src/products/components/ProductUpdatePage/ProductUpdatePage.tsx:191]: I'm not convinced that it's a good name... how about `canOpenAssignReferencesAttributeDialog`?

**Review by dominik-zeglen — APPROVED** [review]

File: src/products/components/ProductUpdatePage/ProductUpdatePage.tsx [hunk 10]
```diff
@@ -69,8 +76,14 @@ export interface ProductUpdatePageProps extends ListActions, ChannelProps {
   saveButtonBarState: ConfirmButtonTransitionState;
   warehouses: WarehouseFragment[];
   taxTypes: TaxTypeFragment[];
+  referencePages: SearchPages_search_edges_node[];
+  assignReferencesAttributeId?: string;
+  fetchMoreReferencePages?: FetchMoreProps;
   fetchCategories: (query: string) => void;
   fetchCollections: (query: string) => void;
+  fetchReferncePages?: (data: string) => void;
```
[the remark this answers is not in the document]
Comment (orzechdev) [reply src/products/components/ProductUpdatePage/ProductUpdatePage.tsx]: My Mac keyboard sometimes doesn't like that character  😞

File: src/products/components/ProductUpdatePage/ProductUpdatePage.tsx (hunk 9 above)
Comment (orzechdev) [reply src/products/components/ProductUpdatePage/ProductUpdatePage.tsx]: fixed

**Review by mmarkusik — APPROVED** [review]

**Pull request merged**